README
Welcome to AnyDBM.

Please visit the homepage at

  http://software.complete.org/anydbm



debian/changelog
haskell-anydbm (1.0.5.4) unstable; urgency=high

  * Rebuild with newer GHC.

 -- John Goerzen <jgoerzen@complete.org>  Mon, 19 Jan 2009 15:53:01 -0600

haskell-anydbm (1.0.5.3) unstable; urgency=low

  * Rebyuild with newer MissingH.

 -- John Goerzen <jgoerzen@complete.org>  Thu, 11 Dec 2008 09:13:53 -0600

haskell-anydbm (1.0.5.2) unstable; urgency=low

  * Rebuild for newer MissingH.
  * Introduces dh_haskell_depends.

 -- John Goerzen <jgoerzen@complete.org>  Fri, 05 Dec 2008 15:38:43 -0600

haskell-anydbm (1.0.5.1) unstable; urgency=low

  * Rebuild for newer MissingH.  Closes: #478939, #477007.

 -- John Goerzen <jgoerzen@complete.org>  Fri, 09 May 2008 09:33:55 -0500

haskell-anydbm (1.0.5.0) unstable; urgency=low

  * Rebuild for newer GHC6, MissingH.

 -- John Goerzen <jgoerzen@complete.org>  Wed, 16 Jan 2008 10:39:21 -0600

haskell-anydbm (1.0.4) unstable; urgency=medium

  * Rebuild for newer MissingH.  Closes: #429707, #431857.
  * Drop Hugs support due to lack of regex.

 -- John Goerzen <jgoerzen@complete.org>  Thu, 12 Jul 2007 00:09:24 -0500

haskell-anydbm (1.0.3) unstable; urgency=low

  * Rebuilt for GHC 6.6.
  * Added build-dep on libhugs-hunit.  Closes: #422628.

 -- John Goerzen <jgoerzen@complete.org>  Fri, 01 Jun 2007 22:09:07 -0500

haskell-anydbm (1.0.2) unstable; urgency=low

  * Rebuild against newer MissingH.
  * Update for newer Hugs.

 -- John Goerzen <jgoerzen@complete.org>  Fri, 04 May 2007 03:48:38 -0500

haskell-anydbm (1.0.1) unstable; urgency=low

  * Switched to correct Setup.hs.

 -- John Goerzen <jgoerzen@complete.org>  Thu,  7 Dec 2006 18:07:34 -0600

haskell-anydbm (1.0.0) unstable; urgency=low

  * Initial release of AnyDBM.  Closes: #402078.
    This package was split off from MissingH due to the MissingH
    transition plan.

 -- John Goerzen <jgoerzen@complete.org>  Thu,  7 Dec 2006 06:46:41 -0600

missingh (0.18.0) unstable; urgency=low

  * Major API changes in this release; several large modules being
    split off into separate packages.  See 
    http://software.complete.org/missingh/wiki/TransitionPlanning
    for details.

 -- John Goerzen <jgoerzen@complete.org>  Thu,  7 Dec 2006 06:13:58 -0600

missingh (0.16.3) unstable; urgency=low

  * New version
  * New module MissingH.ProgressTracker which tracks the progress
    of long-running tasks.  It can also give estimated time to completion
    and overall speed statistics, as well as support dividing a task
    into many subtasks.
  * New module MissingH.ProgressMeter which displays one or more
    ProgressTrackers as a status bar on a console.
  * New module MissingH.Quantity which renders numbers according
    to a quantification system such as SI or binary
  * New functions forkRawSystem and pOpen3Raw in MissingH.Cmd.  Instead
    of taking functions to run as actions, instead returns the PID
    of the child process and leaves it up to the programmer to properly
    handle the result.

 -- John Goerzen <jgoerzen@complete.org>  Tue, 28 Nov 2006 05:52:08 -0600

missingh (0.16.2) unstable; urgency=low

  * Documented Quantity.hs
  * Fixed missing libghc6-*-dev deps from GHC 6.6.  Closes: #394580.

 -- John Goerzen <jgoerzen@complete.org>  Sat, 21 Oct 2006 19:51:08 -0500

missingh (0.16.1) unstable; urgency=low

  * Fix up dependencies

 -- John Goerzen <jgoerzen@complete.org>  Thu, 19 Oct 2006 05:24:04 -0500

missingh (0.16.0) unstable; urgency=low

  * NEW: MissingH.Quantity that is used to render numbers as binary
    units (MB, KB, etc) or SI units, or other similar units
  * NEW: MissingH.ProgressTracker that is used to track the progress
    of long-running tasks, measure their speed, and give estimated
    completion times
  * Update to support GHC 6.6
    + Remove FiniteMap support everywhere (no longer supported in GHC)
    + Remove MissingH.Printf (has been deprecated since 0.13.0 due to
      Text.Printf)
    + Update README
    + Fix MissingH.FTP.Server for GHC 6.6.0 compatibility
    + Update cabal file for GHC 6.6 compatibility
  * Patch from Jeremy Shaw to add to MissingH.HUnit the ability to turn
    QuickCheck tests into HUnit tests
  * Patch from Joe Edmonds to MissingH.Wash.Mail.MailParser to
    accept a slight violation of RFC2822 in which date headers may have
    a single-digit date
  * Fix a logic error in MemoryBuffer's vPutStr that could corrupt
    data when a write occurs after a seek, and wrote a test to
    detect this condition.  Thanks to Bulat Ziganshin for finding this.
  * New function in MissingH.List to merge two sorted lists into
    a single sorted whole.  Patch from Clifford Beshers.
  * Work around lack of signals on Windows in MissingH.Cmd.
  * Make sure Data.Bits is imported in MissingH.IO.StatCompat.

 -- John Goerzen <jgoerzen@complete.org>  Thu, 19 Oct 2006 04:38:10 -0500

missingh (0.14.5) unstable; urgency=low

  * New function: MissingH.Cmd.posixRawSystem
  * MissingH.Cmd.safeSystem now implemented in terms of posixRawSystem 
    on supported platforms.

 -- John Goerzen <jgoerzen@complete.org>  Tue,  4 Jul 2006 11:34:49 -0500

missingh (0.14.4) unstable; urgency=low

  * Fix a bug in ConfigParser.merge and write a test for it.
  * Rebuilt with GHC 6.4.2

 -- John Goerzen <jgoerzen@complete.org>  Wed, 28 Jun 2006 14:05:56 -0500

missingh (0.14.3) unstable; urgency=low

  * IO/StatCompat: Fixed missing Data.Bits import on mingw32 
  * Updated COPYING file with new FSF address
  * Cmd.hs: Commented out code that was never used
  * Import new package MD5 0.2.7 from Ian Lynagh

 -- John Goerzen <jgoerzen@complete.org>  Wed, 28 Jun 2006 03:48:24 -0500

missingh (0.14.2) unstable; urgency=low

  * ConfigParser gives more helpful error messages when possible --
    it now includes section and option in most error messages
  * Unit tests updated for new ConfigParser output
  * MissingH.Str: New function escapeRe that takes a string and makes
    a regular expression that matches the string literally.

testsrc/Tests.hs
{- arch-tag: Tests main file
Copyright (C) 2004 John Goerzen <jgoerzen@complete.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
-}

module Tests(tests) where
import Test.HUnit
import qualified AnyDBMtest

test1 = TestCase ("x" @=? "x")

tests = TestList [TestLabel "test1" test1,
                 TestLabel "AnyDBM" AnyDBMtest.tests]




src/Database/AnyDBM/MapDBM.hs
{-
Copyright (C) 2005 John Goerzen <jgoerzen@complete.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
-}

{- |
   Module     : Database.AnyDBM.MapDBM
   Copyright  : Copyright (C) 2005 John Goerzen
   License    : GNU LGPL, version 2.1 or above

   Maintainer : John Goerzen <jgoerzen@complete.org>
   Stability  : provisional
   Portability: portable

Written by John Goerzen, jgoerzen\@complete.org

Support for working with Maps through the "Database.AnyDBM" framework.
-}

module Database.AnyDBM.MapDBM (MapDBM,
                               newMapDBM,
                               setMapDBM,
                               getMapDBM
                                    )
where
import Database.AnyDBM
import Data.Map as Map
import Control.Concurrent.MVar

{- | The type of the MapDBM. -}
type MapDBM = MVar (Map.Map String String)

{- | Makes a new MapDBM with an empty Map. -}
newMapDBM :: IO MapDBM
newMapDBM = newMVar Map.empty

{- | Sets the embedded Map in this 'MapDBM' to the
given 'Map'. -}
setMapDBM :: MapDBM -> Map.Map String String -> IO ()
setMapDBM h fm = swapMVar h fm >> return ()

{- | Gets the embedded Map in this 'MapDBM'. -}
getMapDBM :: MapDBM -> IO (Map.Map String String)
getMapDBM = readMVar

m = modifyMVar_

instance AnyDBM MapDBM where
    insertA h k v = m h (\x -> return $ Map.insert k v x)
    deleteA h k = m h (\x -> return $ Map.delete k x)
    lookupA h k = withMVar h (\x -> return $ Map.lookup k x)
    toListA h = withMVar h (\x -> return $ Map.toList x)


src/Database/AnyDBM/StringDBM.hs
{- arch-tag: Support for persistent storage through strings
Copyright (C) 2005 John Goerzen <jgoerzen@complete.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
-}

{- |
   Module     : Database.AnyDBM.StringDBM
   Copyright  : Copyright (C) 2005 John Goerzen
   License    : GNU LGPL, version 2.1 or above

   Maintainer : John Goerzen <jgoerzen@complete.org>
   Stability  : provisional
   Portability: portable

Written by John Goerzen, jgoerzen\@complete.org

This 'Database.AnyDBM.AnyDBM' implementation is very simple.  It can store
data on-disk in a persistent fashion, using a very simple String
representation.  While the file is open, an in-memory cache is maintained.
The data is written out during a call to 'flush' or 'close'.
-}

module Database.AnyDBM.StringDBM (StringDBM,
                                  openStringDBM,
                                  openStringVDBM,
                                  SystemFS(..),
                                  IOMode(..)
                                 )
where
import Database.AnyDBM
import System.IO
import System.IO.HVFS
import System.IO.HVIO
import Data.HashTable
import Control.Concurrent.MVar

{- | The type of the StringDBM instances. -}
data StringDBM = forall a. HVFSOpenable a => StringDBM (MVar ()) (HashTable String String) IOMode a FilePath

{- | Opens a 'StringDBM' file.  Please note: only ReadMode, WriteMode,
and ReadWriteMode are supported for the IOMode.  AppendMode is not supported. 

>openStringDBM = openStringVDBM SystemFS
-}
openStringDBM :: FilePath -> IOMode -> IO StringDBM
openStringDBM = openStringVDBM SystemFS

{- | Opens a 'StringDBM' file.  Please note: only ReadMode, WriteMode,
and ReadWriteMode are supported for the IOMode.  AppendMode is not supported.

To work on your system's normal (real) filesystem, just specify
'SystemFS' for the first argument.
-}
openStringVDBM :: HVFSOpenable a => a -> FilePath -> IOMode -> IO StringDBM
openStringVDBM _ _ AppendMode = fail "openStringDBM: AppendMode is not supported"
openStringVDBM h fp ReadMode =
    do ht <- new (==) hashString
       lock <- newMVar ()
       vReadFile h fp >>= strToA ht
       return $ StringDBM lock ht ReadMode h fp
openStringVDBM h fp WriteMode =
    do ht <- new (==) hashString
       lock <- newMVar ()
       return $ StringDBM lock ht WriteMode h fp
openStringVDBM h fp ReadWriteMode =
    -- Nothing different to start with.  Later, we emulate WriteMode.
    -- Nothing is ever read after the object is created.
    do ht <- new (==) hashString
       lock <- newMVar ()
       d <- vDoesFileExist h fp
       if d
          then vReadFile h fp >>= strToA ht
          else return ()
       return $ StringDBM lock ht WriteMode h fp

g :: StringDBM -> HashTable String String
g (StringDBM _ ht _ _ _) = ht

instance AnyDBM StringDBM where
    flushA (StringDBM lock ht WriteMode h fp) = withMVar lock $ \_ ->
        do s <- strFromA ht
           vWriteFile h fp s
    flushA _ = return ()

    insertA = insertA . g
    deleteA = deleteA . g
    lookupA = lookupA . g
    toListA = toListA . g


src/Database/AnyDBM.hs
{- arch-tag: Generic Dict-Like Object Support
Copyright (C) 2005 John Goerzen <jgoerzen@complete.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
-}

{- |
   Module     : Database.AnyDBM
   Copyright  : Copyright (C) 2005 John Goerzen
   License    : GNU LGPL, version 2.1 or above

   Maintainer : John Goerzen <jgoerzen@complete.org>
   Stability  : provisional
   Portability: portable

Written by John Goerzen, jgoerzen\@complete.org

This module provides a generic infrastructure for supporting storage of
hash-like items with String -> String mappings.  It can be used for in-memory
or on-disk items.

-}
module Database.AnyDBM (-- * The AnyDBM class
                        AnyDBM(..),
                        -- * AnyDBM utilities
                        mapA,
                        strFromA, strToA
                       )
where
import Prelude hiding (lookup)
import System.IO
import Data.HashTable
import Control.Exception

{- | The main class for items implementing this interface.

People implementing this class should provide methods for:

* 'closeA' (unless you have no persistent storage)

* 'flushA' (unless you have no persistent storage)

* 'insertA'

* 'deleteA'

* 'lookupA'

* either 'toListA' or 'keysA'
-}
class AnyDBM a where
    {- | Close the object, writing out any unsaved data to disk if necessary.

         If you implement this, make sure your implementation calls 'flushA'.

         Note: if you have an object opened for writing, you MUST
         call closeA on it when you are done.  Implementations are not
         required to preserve your data otherwise.
       -}
    closeA :: a -> IO ()

    {- | Flush the object, saving any un-saved data to disk but not closing
         it. Called automatically by 'closeA'. -}
    flushA :: a -> IO ()

    {- | Insert the given data into the map. Existing data with the same key
       will be overwritten. -}
    insertA :: a             -- ^ AnyDBM object
            -> String           -- ^ Key
            -> String        -- ^ Value
               -> IO ()

    {- | Delete the data referenced by the given key.  It is not an error
         if the key does not exist. -}
    deleteA :: a -> String -> IO ()

    {- | True if the given key is present. -}
    hasKeyA :: a -> String -> IO Bool

    {- | Find the data referenced by the given key. -}
    lookupA :: a -> String -> IO (Maybe String)

    {- | Look up the data and raise an exception if the key does not exist.
         The exception raised is PatternMatchFail, and the string accompanying
         it is the key that was looked up.-}
    forceLookupA :: a -> String -> IO String

    {- | Call 'insertA' on each pair in the given association list, adding
       them to the map. -}
    insertListA :: a -> [(String, String)] -> IO ()

    {- | Return a representation of the content of the map as a list. -}
    toListA :: a -> IO [(String, String)]

    {- | Returns a list of keys in the 'AnyDBM' object. -}
    keysA :: a -> IO [String]

    {- | Returns a list of values in the 'AnyDBM' object. -}
    valuesA :: a -> IO [String]

    valuesA h = do l <- toListA h
                   return $ map snd l

    keysA h = do l <- toListA h
                 return $ map fst l


    toListA h = 
        let conv k = do v <- forceLookupA h k
                        return (k, v)
            in do k <- keysA h
                  mapM conv k

    forceLookupA h key = 
        do x <- lookupA h key
           case x of 
                  Just y -> return y
                  Nothing -> throwIO $ PatternMatchFail key
        
    insertListA h [] = return ()
    insertListA h ((key, val):xs) = do insertA h key val
                                       insertListA h xs

    hasKeyA h k = do l <- lookupA h k
                     case l of
                            Nothing -> return False
                            Just _ -> return True

    closeA h = flushA h

    flushA h = return ()
                  
{- | Similar to MapM, but for 'AnyDBM' objects. -}
mapA :: AnyDBM a => a -> ((String, String) -> IO b) -> IO [b]
mapA h func = do l <- toListA h
                 mapM func l

{- | Similar to 'Data.List.Utils.strToAL' -- load a string representation
into the AnyDBM.  You must supply an existing AnyDBM object;
the items loaded from the string will be added to it. -}
strToA :: AnyDBM a => a -> String -> IO ()
strToA h s = insertListA h (strToAL s)

{- | Similar to 'Data.List.Utils.strFromAL' -- get a string representation of
the entire AnyDBM. -}
strFromA :: AnyDBM a => a -> IO String
strFromA h = do l <- toListA h 
                return (strFromAL l)

instance AnyDBM (HashTable String String) where
    insertA h k v = do delete h k
                       insert h k v
    deleteA = delete
    lookupA = lookup
    toListA = toList
    

{- The following functions imported from MissingH to avoid a dep on
 - MissingH -}
strToAL :: (Read a, Read b) => String -> [(a, b)]
strToAL inp =
    let worker line =
            case reads line of
               [(key, remainder)] -> case remainder of
                     ',':valstr -> (key, read valstr)
                     _ -> error "Data.List.Utils.strToAL: Parse error on value"
               _ -> error "Data.List.Utils.strToAL: Parse error on key"
        in map worker (lines inp)

strFromAL :: (Show a, Show b) => [(a, b)] -> String
strFromAL inp =
    let worker (key, val) = show key ++ "," ++ show val
        in unlines . map worker $ inp

